autoreload/autoreload.py

```diff
@@ -0,0 +1,227 @@
+            except ValueError:
+                # Probably didn't read enough
+                buff = buff.lstrip().encode("UTF-8")
+
+    def start(self):
+        assert(self.status == 'stopped')
+        try:
+            self.proc = subprocess.Popen([self.path], stdout=subprocess.PIPE, stdin=subprocess.PIPE)
+            self.status = 'started'
+            self.getmanifest()
+            return True
+        except:
+            return False
+
+    def stop(self):
+        assert(self.status == 'started')
+        self.proc.kill()
+        self.proc.wait()
+        reader = self.reader
+        if reader:
+            reader.join()
+        self.status = 'stopped'
+
+    def restart(self):
+        print('Restarting child plugin')
+        self.stop()
+        self.start()
+        plugin.child.send(self.init)
+        print("Sent init to child plugin")
+        plugin.child.passthru()
+
+    def getmanifest(self):
+        assert(self.status == 'started')
+        self.send({'jsonrpc': '2.0', 'id': 0, 'method': 'getmanifest', 'params': []})
+
+        while True:
+            msg = self._readobj(self.proc.stdout)
+            if 'id' in msg and msg['id'] == 0:
+                self.manifest = msg['result']
+                break
+            self.plugin._write_locked(msg)
+        return self.manifest
+
+    def passthru(self):
+
+        # First read the init reply, and then we can switch to passthru
+        while True:
+            msg = self._readobj(self.proc.stdout)
+            if 'id' in msg and msg['id'] == self.init['id']:
+                break
+            self.plugin._write_locked(msg)
+
+        def read_loop():
+            while True:
+                line = self.proc.stdout.readline()
+                if line == b'':
+                    break
+                self.plugin.stdout.buffer.write(line)
+                self.plugin.stdout.flush()
+            self.reader = None
+            print("Child plugin exited")
+        self.reader = threading.Thread(target=read_loop)
+        self.reader.daemon = True
+        self.reader.start()
+
+    def send(self, msg):
+        self.proc.stdin.write(json.dumps(msg).encode('UTF-8'))
+        self.proc.stdin.write(b'\n\n')
+        self.proc.stdin.flush()
+
+    def proxy_method(self, request, *args, **kwargs):
+        raw = {
+            'jsonrpc': '2.0',
+            'method': request.method,
+            'params': request.params,
+            'id': request.id,
+        }
+        self.send(raw)
+
+    def proxy_subscription(self, request, *args, **kwargs):
+        raw = {
+            'jsonrpc': '2.0',
+            'method': request.method,
+            'params': request.params,
+        }
+        self.send(raw)
+
+
+@plugin.init()
+def init(options, configuration, plugin, request):
+    if options['autoreload-plugin'] in ['null', None]:
+        print("Cannot run the autoreload plugin on its own, please specify --autoreload-plugin")
+        plugin.rpc.stop()
```

Comment:
I think you need to return here and not fall thru and start the watch_thread below, right?

Reply:
Good point, by the time `stop()` returns we're already dead, but always good to make sure we don't do useless work.